lib/CodeGen/CGExprScalar.cpp

```diff
@@ -2427,6 +2433,9 @@ ScalarExprEmitter::EmitScalarPrePostIncDec(const UnaryOperator *E, LValue LV,
 
   // Next most common: pointer increment.
   } else if (const PointerType *ptr = type->getAs<PointerType>()) {
+    // Insert a dynamic check for arithmetic on null checked pointers.
```

Comment:
I am not sure about adding a compile time check to see if `value` is indeed null so that the runtime check can be elided. LLVM has a function called `isKnownNonNull` which is used in a couple of places in Clang but in my case it always returns 0.

Consider this code:
```
array_ptr<char> s = NULL;
++s;
```

This is the IR:
```
 %s = alloca i8*, align 8
 store i8* null, i8** %s, align 8
 %0 = load i8*, i8** %s, align 8
 %_Dynamic_check.non_null = icmp ne i8* %0, null
```

I guess we need to check what is getting stored in `%s`. If it is indeed non-null, then we can elide the check. Is this a valid approach?